src/Microsoft.ML.Recommender/MatrixFactorizationPredictor.cs

```diff
@@ -0,0 +1,496 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.ML.Runtime;
+using Microsoft.ML.Runtime.Data;
+using Microsoft.ML.Runtime.Internal.Internallearn;
+using Microsoft.ML.Runtime.Internal.Utilities;
+using Microsoft.ML.Runtime.Model;
+using Microsoft.ML.Runtime.Recommender;
+using Microsoft.ML.Runtime.Recommender.Internal;
+using Microsoft.ML.Runtime.Model.Pfa;
+using Microsoft.ML.Runtime.Model.Onnx;
+using Microsoft.ML.Runtime.CommandLine;
+using Microsoft.ML.Runtime.Data.IO;
+
+[assembly: LoadableClass(typeof(MatrixFactorizationPredictor), null, typeof(SignatureLoadModel), "Matrix Factorization Predictor Executor", MatrixFactorizationPredictor.LoaderSignature)]
+
+[assembly: LoadableClass(typeof(MatrixFactorizationPredictionTransformer), typeof(MatrixFactorizationPredictionTransformer),
+    null, typeof(SignatureLoadModel), "", MatrixFactorizationPredictionTransformer.LoaderSignature)]
+
+namespace Microsoft.ML.Runtime.Recommender
+{
+    public sealed class MatrixFactorizationPredictor : IPredictor, ICanSaveModel, ICanSaveInTextFormat, ISchemaBindableMapper
+    {
+        internal const string LoaderSignature = "MFPredictor";
+        internal const string RegistrationName = "MatrixFactorizationPredictor";
+
+        private static VersionInfo GetVersionInfo()
+        {
+            return new VersionInfo(
+                modelSignature: "FAFAMAPD",
+                verWrittenCur: 0x00010001,
+                verReadableCur: 0x00010001,
+                verWeCanReadBack: 0x00010001,
+                loaderSignature: LoaderSignature,
+                loaderAssemblyName: typeof(MatrixFactorizationPredictor).Assembly.FullName);
+        }
+
+        private readonly IHost _host;
+        // The number of rows.
+        private readonly int _m;
+        // The number of columns.
+        private readonly int _n;
+        // The internal dimension.
+        private readonly int _k;
```

Comment:
_internalDimension?

Reply:
For mathmatical notations, I prefer to use the same notations used across my papers. Otherwise, it'd be very difficult to debug my code.

Reply:
Ok, here is my problems.
A) What paper? It's nowhere mentioned in PR, only links I can find is in that submodule git.
B) Predictor doesn't even touch libMF native code.
C) This is open source code,  it's not your code anymore, and you basically force person to read your paper (which is impossible to find) to understand what is going on here. I can jump back and forth to comments, by why should I do that? 

You can name them in a way person who wouldn't read you paper would understand what they for, and put in comments your reference to paper instead.



---
In reply to: [227611326](https://github.com/dotnet/machinelearning/pull/1263#discussion_r227611326) [](ancestors = 227611326)

Reply:
They might not understand those variables without reading papers. For example, matrix factorization trainer seeks for two factor matrices to approximate the training matrix. Assume that the two matrices found by trainer is P and Q. What is the approximation formulation? Is it PQ^T, Q^TP, PQ or some magic formulations used in my work? It's clearly a LIBMF submodule. User should go to LIBMF pages if more information is needed.

---
In reply to: [227890479](https://github.com/dotnet/machinelearning/pull/1263#discussion_r227890479) [](ancestors = 227890479,227611326)